Detect unneeded `async` keywords on functions

plredmond opened the pull request:
## Summary

This change adds a rule to detect functions declared `async` but lacking any of `await`, `async with`, or `async for`. This resolves #9951.

## Test Plan

This change was tested by following https://docs.astral.sh/ruff/contributing/#rule-testing-fixtures-and-snapshots and adding positive and negative cases for each of `await` vs nothing, `async with` vs `with`, and `async for` vs `for`.

## Questions/todo

* [x] I'm unsure whether the rule should also check for `yield` appearing in functions. My understanding is that there has been several rounds of iteration in how python async is supposed to be done, and that `yield` is not considered current, and so I haven't added a case for that. Therefore an `async` function containing only `yield` will cause this rule to fire.
  * resolved by "async doesn't require yield" so the rule will fire when an async function's only async feature is yield
  * https://github.com/astral-sh/ruff/pull/9966#issuecomment-2044527147
  * https://github.com/astral-sh/ruff/pull/9966#pullrequestreview-2002255367
* [x] I was unsure how to access just the range of the function declaration (without the body), so the rule currently indicates the whole function declaration & body as the problem area.
* ~~[ ] Needs review by @zanieb~~

plredmond commented:
Sorry, I neglected formatting and linting of the rust code. I'll do that now.

zanieb commented on crates/ruff_linter/resources/test/fixtures/ruff/RUF029.py:
```diff
@@ -0,0 +1,30 @@
+import time
+import asyncio
+
+async def pass_case(): # passes because it awaits something
```
Usually we'd indicate this with something like 

```suggestion
async def pass_case(): # OK: awaits a coroutine
```

zanieb replied:
It's not really standardized though, so it's okay either way.

plredmond replied:
I've made this change below.

zanieb commented:
The `rasa` ecosystem checks look like a bunch of false positives due to methods that are overriding an [abstract method which must be async](https://github.com/RasaHQ/rasa/blob/cca30d4e06af5aba177e916d64c60313fc537005/rasa/core/actions/action.py#L241-L263). This common enough that we should attempt to avoid it, but it requires multifile analysis in most cases which we do not yet support yet. There are some other issues like this, but I can't recall them — perhaps @charliermarsh knows. We could consider just not applying this to methods in classes with base classes for now?

charliermarsh commented:
Ah yeah, the thing we often do there (at the very least) is check if the method has an `@override` decorator (can grep for `is_override`), since that's used to hint to static analysis tools that the method doesn't have control over its own signature. So users at least have a way to opt-out of these kinds of rules entirely for methods that override a parent method. I would be fine omitting this entirely for classes with base classes though (or even classes at all?).

zanieb commented:
Here's another interesting edge-case false positive https://github.com/zulip/zulip/blob/35098f49597895718343091881fbd6198bd2022d/zerver/tornado/views.py#L35 — this one I'm less sure we can/should do anything about.

zanieb commented on crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs:
```diff
@@ -0,0 +1,83 @@
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for functions declared `async` that do not yield or otherwise use features requiring the
+/// function to be declared `async`.
+///
+/// ## Why is this bad?
+/// Declaring a function `async` when it's not is usually a mistake, and may artificially limit the
+/// contexts where that function may be called.
```
I think we'll want to include an example here too

zanieb replied:
You can also probably say "will limit the contexts" instead of "may artificially limit".

There's also something to be said about it actually being _bad_ to say that your function is async without it containing a yield as it's breaking the fundamental promise of the async label. Some more context on this idea [in the trio documentation](https://trio.readthedocs.io/en/stable/reference-core.html#checkpoints). I'm not sure if we can get into that without confusing users though.

plredmond replied:
I've made this change below.

charliermarsh commented on crates/ruff_linter/resources/test/fixtures/ruff/RUF029.py:
```diff
@@ -0,0 +1,30 @@
+
+async def fail_case_3(): # RUF029
+    for i in []:
+        pass
```
Do you mind running `ruff format` over this file, just to standardize the comment spacing, lines between functions, etc.? (We don't _always_ run it over fixtures, since fixtures sometimes include intentionally-odd comments and syntax, but this one doesn't, so seems like a nice improvement.)

plredmond replied:
Formatted!

charliermarsh commented on crates/ruff_linter/src/checkers/ast/analyze/statement.rs:
```diff
@@ -366,6 +366,9 @@ pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
                 flake8_bandit::rules::ssl_with_bad_defaults(checker, function_def);
             }
+            if checker.enabled(Rule::SpuriousAsync) {
+                ruff::rules::spurious_async(checker, *is_async, name, body, stmt.range());
```
For newer rules, we're trying to pass the entire node (like `function_def`, above), to ensure that callers can't pass in "incorrect" data. Do you mind revising the signature?

plredmond replied:
I've made this change below.

charliermarsh reviewed (commented):
Nice!

zanieb commented:
Perhaps @AlexWaygood can elucidate whether or not this should trigger for functions with `yield`

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    let yields = {
+        let mut visitor = YieldingExprVisitor { found_yield: false };
```
Nit: You could derive `Default` for `YieldingExprVisitor` so that this becomes
```suggestion
        let mut visitor = YieldingExprVisitor::default();
```

plredmond replied:
Yes, I followed that pattern in an earlier draft when the field was a vector of yielding-expression-TextRanges, but reverted to explicit construction when I changed it to a boolean. Is the convention to always prefer default where applicable?

MichaReiser replied:
There's no strong convention for private types. For public types, yes, I would say we prefer using `::default` or `::new` to encapsulate the initialization logic. Here I would do it because it's less to type but that's also why it is a Nit. It's up to you what you prefer.

plredmond replied:
I've made this change below.

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs:
```diff
@@ -0,0 +1,83 @@
+    body: &[Stmt],
+    range: TextRange,
+) {
+    if !is_async {
```
This is a Python noob question from a JS engineer. How does Python handle default arguments? Can you use `await` in a default initializer? Does this count as async call (because it pauses the function execution) or not because defaults are initialized once during declaration and not when calling the function.

AlexWaygood replied:
That's a fun question that I'd never considered before.

`await` isn't allowed at the top-level in a usual Python file (only inside async functions), so normally this isn't really something that would ever come up, since function defaults are evaluated in the scope the function is in rather than the scope the function creates:

```pycon
>>> import asyncio
>>> async def foo(x=await asyncio.sleep(0.1)): ...
...
  File "<stdin>", line 1
SyntaxError: 'await' outside function
```

But theoretically you could use `await` in a function default argument if it was an inner function inside an async function:

```pycon
>>> import asyncio
>>> async def foo():
...     def bar(x=await asyncio.sleep(0.1)): ...
...     return bar
...
>>> bar = asyncio.run(foo())
>>> import inspect
>>> inspect.iscoroutinefunction(bar)
False
```

As you can see from that REPL snippet, using `await` inside the default argument for the inner function doesn't impact whether or not the inner function is considered to be a coroutine function, since the `await` in the function's default argument is evaluated before the function has even finished being defined.

MichaReiser replied:
Thanks for the detailed explanation.  

I tested if `foo` is considered a co-routine and that seems to be the case. Do you know if that's only because the function is marked as `async`? 

```python
>>> import asyncio
>>> async def foo():
...     def bar(x=await asyncio.sleep(0.1)): ...
...     return bar
... 
>>> import inspect
>>> inspect.iscoroutinefunction(foo)
True
```

I'm trying to understand if the use of `async` on `foo` is something this lint rule should warn about or not.

AlexWaygood replied:
`foo` is considered a coroutine function because it returns a coroutine ;)

```pycon
>>> foo
<function foo at 0x00000274CBC2A160>
>>> foo()
<coroutine object foo at 0x00000274CDC19300>
```

But the short answer to your question is "yes" -- all `async def` functions are "coroutine functions" -- functions that return coroutine objects that can then be `await`ed inside other coroutine functions.

The use of `async` on `foo` should not be something that this lint rule should warn about, because the `await` expression in the default for the `x` parameter of `bar` is executed within the scope of `foo`. So removing the `async` on `foo` would mean that that `await` expression would no longer be found in an async context (== illegal syntax).

MichaReiser replied:
Perfect. That's what this lint rule should support already today. It might be worth to copy the example as a test case.

plredmond replied:
I've added a test case (which should pass, without causing the rule to fire) where an async function defines an inner function by setting the inner function's default argument value to the result of an await expression, below.

AlexWaygood commented:
> Perhaps @AlexWaygood can elucidate whether or not this should trigger for functions with `yield`.
> 
> [peps.python.org/pep-0525](https://peps.python.org/pep-0525) may be helpful.

Yeah, async functions that contain `yield` are async generators that can be used in `async for` loops by other async functions. So I don't think ruff should emit this diagnostic for async functions that have a `yield` in them:

```pycon
>>> import asyncio
>>> async def x():
...     yield 42
...     yield 56
...
>>> x
<function x at 0x00000274CDC52200>
>>> x()
<async_generator object x at 0x00000274CC36BE20>
>>> async def y():
...     async for number in x():
...         print(number)
...
>>> asyncio.run(y())
42
56
```

plredmond commented on crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs:
```diff
@@ -0,0 +1,83 @@
+
+#[derive(Default)]
+struct YieldingExprVisitor {
+    found_yield: bool,
```
I originally used "yielding_expression". Does that name seem acceptable?

MichaReiser replied:
Good question. `yield` or `yielding` is too close to the `yield` expression semantics. I would probably be explicit and call it `awaits_or_yields`

plredmond replied:
I've made this rename change below.

MichaReiser commented:
I think there's one issue that we might need to look into and it's that the visitor will cross function boundaries when searching for `await` and `yield`, resulting in false negatives

```python
async def unnecessary_async():
	async def inner():
		await x

async def unnecessary_async2():
	class Inner:
		async def inner():	
			await x


 ```

I don't think the rule flags `unnecessary_async` or `unnecessary_async2`.

plredmond commented:
Thanks for your review! I've looked over the comments above and can work on the changes later today. I need to devote this morning to a different task.

> Ah yeah, the thing we often do there (at the very least) is check if the method has an `@override` decorator (can grep for `is_override`), since that's used to hint to static analysis tools that the method doesn't have control over its own signature. So users at least have a way to opt-out of these kinds of rules entirely for methods that override a parent method. I would be fine omitting this entirely for classes with base classes though (or even classes at all?).

@charliermarsh should I omit this check in both the cases (1) in the context of a class or (2) when the override decorator is present? Based on your description, it seems preferable to limit case (1) to classes with a base class, since they're actually likely to be overriding things, though.

plredmond commented:
> I don't think the rule flags `unnecessary_async` or `unnecessary_async2`.

@MichaReiser These examples make the problem a bit harder. :slightly_smiling_face: I need to check if the visitor has a callback for when an inner-function ends; I could use that to track which scope I'm in when detecting yielding-expressions to attribute them correctly.

plredmond commented on crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs:
```diff
@@ -0,0 +1,83 @@
+    found_yield: bool,
+}
+
+impl<'a> Visitor<'a> for YieldingExprVisitor {
```
Combined with your point about nested-functions (e.g. `unnecessary_async2`) from [this comment](https://github.com/astral-sh/ruff/pull/9966#issuecomment-1942180352), I need to rethink my approach a bit.

plredmond replied:
Implemented this below.

MichaReiser commented:
@plredmond I think there are only two cases where this is relevant: classes and functions. You could extend your visitor to handle `visit_stmt explicitly` and traverse classes and functions manually (don't call `walk_stmt`). This allows you to skip traversing the body (you still want to traverse decorators and the function header)

plredmond commented:
Hi! Sorry to let this PR languish. I have a paper deadline just before I start my astral internship, so I have to put this on ice until then.

MichaReiser commented:
IMO, `unused-async` is easier to understand than spurious (which I mainly associate with threading).

AlexWaygood commented:
> Yeah, async functions that contain `yield` are async generators that can be used in `async for` loops by other async functions. So I don't think ruff should emit this diagnostic for async functions that have a `yield` in them:

So, I chatted to @carljm about this offline... and he persuaded me that the rationale I gave in https://github.com/astral-sh/ruff/pull/9966#issuecomment-1941990446 for excluding async generators from this check _basically_ makes no sense :-)

It's true that removing the `async` keyword from a function with a `yield` statement in it will mean that it will become a sync generator function rather than an async generator function. That means that it will only be usable in `for` loops, not `async for` loops, which means callsites of the function will have to be rewritten to use `for` loops rather than `async for` loops -- potentially a sweeping and disruptive refactor. _However_... that's not really materially different from the changes this rule recommends to non-generator functions. Removing redundant `async` keywords from non-generator functions that don't have any `await`s in them will _also_ necessitate callsites being refactored: you'll have to do this refactor everywhere:

```diff
- x = await function_in_question()
+ x = function_in_question()
```

As such, I'm now persuaded that there's no reason to exclude async generator functions from this check. (Thanks @carljm!)

plredmond commented:
Picking this up today. Sorry for the delay.

plredmond commented:
Saving the rename for last b/c it'll disrupt the above comments on diffs.

plredmond commented on crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs:
```diff
@@ -0,0 +1,83 @@
+            SpuriousAsync {
+                name: name.to_string(),
+            },
+            range,
```
Is it desirable for the error message to only indicate the name, given that the problem is to the left of the name (the use of the `async` keyword) and to the right of it (the function body lacking a reason for the use of the `async` keyword)?

plredmond commented:
I've made a bunch of smaller/cosmetic changes per the comments above. Additionally, I've added four new test cases (two pass, and two fail). The remaining work is:

* make the new test cases pass
* fix the visitor to stop traversing early when it finds a problem
* rename "spurious async" to "unused async" throughout

Could somebody take a look at the two pass-cases and fail-cases that I just added and confirm that this is the desired behavior?

carljm commented on crates/ruff_linter/resources/test/fixtures/ruff/RUF029.py:
```diff
@@ -0,0 +1,56 @@
+    return foo
+
+
+async def pass_case_5():  # OK: yields a value
+    yield "hello"
```
For the reasons mentioned in  @AlexWaygood's most recent comment, this case should fire `RUF029` -- yielding a value does not require an async function.

carljm commented on crates/ruff_linter/resources/test/fixtures/ruff/RUF029.py:
```diff
@@ -0,0 +1,56 @@
+    yield "hello"
+
+
+async def fail_case_6():  # RUF029: the outer function does not await or yield
+    async def foo():
+        await bla
+
+
+async def fail_case_6():  # RUF029: the outer function does not await or yield
+    class Foo:
+        async def foo():
+            await bla
```
and "or yield" should be removed from the comments here

carljm commented on crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs:
```diff
@@ -0,0 +1,98 @@
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let SpuriousAsync { name } = self;
+        format!("Function `{name}` is declared `async`, but doesn't yield or use async features.")
```
and "yield or" should be removed from this error message

carljm reviewed (commented):
The test cases look right to me, apart from the inline comments about `yield`

charliermarsh commented:
Many reviewers on here -- who wants to own the final review + approval? \cc @carljm @AlexWaygood @zanieb @MichaReiser

plredmond commented:
Sorry for the excessive number of commits here. I've had a bit of churn in the test cases because I found them confusing and reorganized them. The real work is

* [switch to preorder visitor](https://github.com/astral-sh/ruff/pull/9966/commits/1cb7765b62046fe32050d08671e08d32a8839a4c)
[1cb7765](https://github.com/astral-sh/ruff/pull/9966/commits/1cb7765b62046fe32050d08671e08d32a8839a4c)
* [skip the rest of traversal after finding an async/await](https://github.com/astral-sh/ruff/pull/9966/commits/41f0ac5bfaaf5ca939be84e003f9fe7d008b7b9c)
[41f0ac5](https://github.com/astral-sh/ruff/pull/9966/commits/41f0ac5bfaaf5ca939be84e003f9fe7d008b7b9c)
* [don't traverse the body of inner-functions or inner-classes](https://github.com/astral-sh/ruff/pull/9966/commits/b714e808d8f403bdecbaa9a16df0f4b548261d33)
[b714e80](https://github.com/astral-sh/ruff/pull/9966/commits/b714e808d8f403bdecbaa9a16df0f4b548261d33)

I believe this covers the test cases @MichaReiser shared [above](https://github.com/astral-sh/ruff/pull/9966#issuecomment-1942180352) with the nested asyncs using his [suggested](https://github.com/astral-sh/ruff/pull/9966#issuecomment-1942437370) solution (partially reimplementing traversal of function defs and class defs), as well as his [suggested](https://github.com/astral-sh/ruff/pull/9966#discussion_r1488203954) optimization to stop traversing once some evidence that the function is actually async has been found.

However, this *doesn't* cover @charliermarsh's [suggestion](https://github.com/astral-sh/ruff/pull/9966#issuecomment-1939953539) to just omit checking methods in classes. I'm not sure how to approach that (and the test case is commented out). The rule is currently written in terms of a function definition, however, to omit the rule for methods, we'd need to know whether that function definition is a class method. I didn't see a way to do that in `analyze/statement.rs`.

plredmond commented:
Ok, I think everything is done except for this:

> However, this _doesn't_ cover @charliermarsh's [suggestion](https://github.com/astral-sh/ruff/pull/9966#issuecomment-1939953539) to just omit checking methods in classes. I'm not sure how to approach that (and the test case is commented out). The rule is currently written in terms of a function definition, however, to omit the rule for methods, we'd need to know whether that function definition is a class method. I didn't see a way to do that in `analyze/statement.rs`.

MichaReiser commented:
@plredmond I'm not entirely sure if that's what you mean but you can test if you're inside a class by using 

```
    // Assignments in class bodies are attributes (e.g., `x = x` assigns `x` to `self.x`, and thus
    // is not a self-assignment).
    if checker.semantic().current_scope().kind.is_class() {
        return;
    }
```

You can also use the `kind` to retrieve more information about the enclosing class (e.g. if it has any base class or whatnot)

kkom commented:
Came across this PR when looking at #9951 – thank you for adding this rule, very excited about it!

Regarding the latest discussion – I'm curious why not omit it for method marked with `@override` and continue linting methods? This is analogous to how https://docs.astral.sh/ruff/rules/no-self-use/ works – and that rule has worked really well for us in that way.

zanieb commented:
@kkom we're just cutting scope from the initial rule. We could consider it in the future.

charliermarsh commented on crates/ruff_linter/src/rules/ruff/rules/unused_async.rs:
```diff
@@ -0,0 +1,169 @@
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let UnusedAsync { name } = self;
+        format!("Function `{name}` is declared `async`, but doesn't await or use async features.")
```
Nit: wrap await in backticks here for consistency with async?

plredmond replied:
Implemented below.

charliermarsh commented:
This looks good to me, but I suggest adding the snippet that Micha posted above.

charliermarsh commented:
@plredmond - Try running `cargo dev generate-all` -- you need to regenerate the JSON Schema since we added a new rule.

charliermarsh commented:
(I believe that's the cause of your failing test.)

plredmond commented:
```sh-session
$ cargo dev generate-all
...
$ git diff
diff --git a/ruff.schema.json b/ruff.schema.json
index 871cce40b..90582fc67 100644
--- a/ruff.schema.json
+++ b/ruff.schema.json
@@ -3951,4 +3951,4 @@
       ]
     }
   }
-}
+}
\ No newline at end of file

charliermarsh commented:
Is your editor adding a newline?

plredmond commented:
I also tried fetching and merging in the latest `main` in. Still no success. Strangely, I'm not seeing this locally.

I'll look at the CI output.

plredmond commented:
> Is your editor adding a newline?

It was modified by `cargo dev generate-all` afaik. My editor is vim and I didn't modify that file.

charliermarsh commented:
We do _expect_ the Schema to change on this branch.

plredmond commented:
I was wrong. I *did* edit it, back in feb, and that seems to have lead to the CI complaining about the newline. It should be fixed now.

charliermarsh commented on crates/ruff_linter/src/codes.rs:
```diff
@@ -962,6 +962,7 @@ pub fn code_to_rule(linter: Linter, code: &str) -> Option<(RuleGroup, Rule)> {
         (Ruff, "025") => (RuleGroup::Preview, rules::ruff::rules::UnnecessaryDictComprehensionForIterable),
         (Ruff, "026") => (RuleGroup::Preview, rules::ruff::rules::DefaultFactoryKwarg),
         (Ruff, "027") => (RuleGroup::Preview, rules::ruff::rules::MissingFStringSyntax),
+        (Ruff, "029") => (RuleGroup::Preview, rules::ruff::rules::UnusedAsync),
```
Sorry, annoying, but do you mind flipping the order here so that this rule comes last (after `RUF028`)?

plredmond replied:
Done
